framework/src/org/checkerframework/common/value/Range.java

```diff
@@ -0,0 +1,742 @@
+            }
+            // If right.to > Long.MIN_VALUE, union the previous result with the result of range
+            // [right.from + 1, right.to] divided by this range, which can be calculated using
+            // the general method (see below)
+            if (right.to > Long.MIN_VALUE) {
+                Range rangeAdditional = this.remainder(new Range(right.from + 1, right.to));
+                range = range.union(rangeAdditional);
+            }
+            return range;
+        }
+        // General method:
+        // Calculate range1: the result range of this range divided by EVERYTHING. For example,
+        // if this range is [3, 5], then the result range would be [0, 5]. If this range is [-3, 4],
+        // then the result range would be [-3, 4]. In general, the result range is {0} union with
+        // this range excluding the value Long.MIN_VALUE.
+        Range range1 =
+                (new Range(Math.max(Long.MIN_VALUE + 1, from), Math.max(Long.MIN_VALUE + 1, to)))
+                        .union(new Range(0, 0));
+        // Calculate range2: the result range of range EVERYTHING divided by the right range. For
+        // example, if the right range is [-5, 3], then the result range would be [-4, 4]. If the
+        // right range is [3, 6], then the result range would be [-5, 5]. In general, the result
+        // range is calculated as following:
+        Range range2 =
+                Math.abs(right.from) > Math.abs(right.to)
+                        ? new Range(-Math.abs(right.from) + 1, Math.abs(right.from) - 1)
+                        : new Range(-Math.abs(right.to) + 1, Math.abs(right.to) - 1);
+        // Since range1 and range2 are both super sets of the minimal result range, we return the
+        // intersect of range1 and range2, which is correct (super set) and precise enough.
+        return range1.intersect(range2);
+    }
+
+    /**
+     * Returns a range that includes all possible values resulting from left shifting an arbitrary
+     * value in this range by an arbitrary number of bits in the specified range. We call this the
+     * left shift operation of a range.
+     *
+     * @param right the range of bits by which this range is left shifted
+     * @return the range resulting from left shifting this range by the specified range
+     */
+    public Range shiftLeft(Range right) {
+        if (this.isNothing() || right.isNothing()) {
+            return NOTHING;
+        }
+
+        if (this.isWithinInteger() && right.from >= 0 && right.to <= 31) {
+            // The long type can handle the cases with shift distance no longer than 31.
+            long resultFrom = from << (from >= 0 ? right.from : right.to);
+            long resultTo = to << (to >= 0 ? right.to : right.from);
+            return new Range(resultFrom, resultTo);
+        } else if (right.from >= 0 && right.to <= 63) {
+            // If the shift distance is within 0 to 63, the calculation can be handled by BigInteger.
```

Comment:
BigInteger can handle any shift distance, can't it?  I think it would be clearer to say that the code chooses to give up if the shift distance is negative or greater than 63.